storage/src/desktop/storage_reference_desktop.cc

```diff
@@ -279,18 +277,96 @@ Future<size_t> StorageReferenceInternal::GetBytes(void* buffer,
+                           controller_out]() -> BlockingResponse* {
+    auto* future_api = future();
+    auto handle =
+        future_api->SafeAlloc<size_t>(kStorageReferenceFnGetBytesInternal);
+    storage::internal::Request* request = new storage::internal::Request();
+    PrepareRequest(request, storageUri_.AsHttpUrl().c_str(), rest::util::kGet);
+    GetBytesResponse* response =
+        new GetBytesResponse(buffer, buffer_size, handle, future_api);
+    RestCall(request, request->notifier(), response, handle.get(), listener,
+             controller_out);
+    return response;
+  }};
+  SendRequestWithRetry(kStorageReferenceFnGetBytesInternal, send_request_funct,
+                       handle, storage_->max_download_retry_time());
+  return GetBytesLastResult();
+}
 
-  GetBytesResponse* response =
-      new GetBytesResponse(buffer, buffer_size, handle, future_api);
+// Sends a rest request, and creates a separate thread to retry failures.
+void StorageReferenceInternal::SendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct, SafeFutureHandle<size_t> final_handle,
+    double max_retry_time_seconds) {
+  BlockingResponse* first_response = send_request_funct();
+  std::thread async_retry_thread(
+      &StorageReferenceInternal::AsyncSendRequestWithRetry, this,
+      internal_function_reference, send_request_funct, final_handle,
+      first_response, max_retry_time_seconds);
+  async_retry_thread.detach();
+}
+
+const int kInitialSleepTimeMillis = 1000;
+const int kMaxSleepTimeMillis = 30000;
+
+// In a separate thread, repeatedly send Rest requests until one succeeds or a
+// maximum amount of time has passed.
+void StorageReferenceInternal::AsyncSendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct, SafeFutureHandle<size_t> final_handle,
+    BlockingResponse* response, double max_retry_time_seconds) {
+  auto* future_api = future();
+  firebase::FutureBase internal_future;
+  auto end_time = std::chrono::steady_clock::now() +
+                  std::chrono::duration<double>(max_retry_time_seconds);
+  auto current_sleep_time = std::chrono::milliseconds(kInitialSleepTimeMillis);
+  while (true) {
+    internal_future = future_api->LastResult(internal_function_reference);
+    // Wait for completion, then check status and error.
+    while (internal_future.status() == firebase::kFutureStatusPending) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+    // For any request that succeeds or fails in a non-retryable way, don't
+    // bother retrying.
+    if (internal_future.status() != firebase::kFutureStatusComplete ||
+        !IsRetryableFailure(response->status())) {
+      break;
+    }
+    // Interrupt the loop if the retry deadline has been reached
+    auto current_time = std::chrono::steady_clock::now();
+    if (current_time + current_sleep_time > end_time) {
```

Comment:
It might be nice, if current_time + current_sleep_time > end_time, but current_sleep_time is quite large, to maybe give it one last retry when the maximum time is almost up?

e.g. Right now, current_sleep_time is 30 seconds, and end_time is 25 seconds from now, it won't actually do a last retry before the 25 seconds are up. I'm not sure if this is a major issue, but it might be nice to detect this case and maybe reduce current_sleep_time a bit.

(Just a suggestion, feel free to ignore.)

Reply:
I based this exponential backoff logic on the ExponentialBackoffSender that is used for firebase android downloads. It has the following check: 
    while (clock.elapsedRealtime() + currentSleepTime <= deadLine ...
so I prefer to keep this logic simpler for now and similarly stop if the sleep time would exceed the deadline.